pkg/kube-metrics/metrics.go

```diff
@@ -0,0 +1,116 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	kcollector "k8s.io/kube-state-metrics/pkg/collector"
+	ksmetric "k8s.io/kube-state-metrics/pkg/metric"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("kubemetrics")
+
+// ServeCRMetrics generates CR specific metrics based on the operator GVK.
+// It starts serving collections of those metrics on given host and port.
+func ServeCRMetrics(cfg *rest.Config, ns []string, operatorTypes map[schema.GroupVersionKind]reflect.Type, host string, port int32) error {
+	// Create new unstructured client.
+	uc := NewClientForConfig(cfg)
+	var collectors [][]*kcollector.Collector
+	log.V(1).Info("Starting collecting operator types")
+
+	// Loop through all the possible operator specific types.
+	for gvk, _ := range operatorTypes {
+		// Generate metric based on the kind.
+		metricFamilies := generateMetricFamilies(gvk.Kind)
+		log.V(1).Info("Generating metric families for kind ", gvk.Kind, "and group/version", gvk.GroupVersion().String())
+		// Generate collector based on the resource, kind and the metric families.
+		c, err := NewCollectors(uc, ns, gvk.GroupVersion().String(), gvk.Kind, metricFamilies)
+		if err != nil {
+			if err == k8sutil.ErrNoNamespace {
+				log.Info("Skipping operator specific metrics; not running in a cluster.")
+				return nil
+			}
+			return err
+		}
+		collectors = append(collectors, c)
+	}
+	// Start serving metrics.
+	log.V(1).Info("Starting serving metric families")
+	go ServeMetrics(collectors, host, port)
+
+	return nil
+}
+
+func generateMetricFamilies(kind string) []ksmetric.FamilyGenerator {
+	helpText := fmt.Sprintf("Information about the %s operator replica.", kind)
+	kindName := fmt.Sprintf("%s", strings.ToLower(kind))
+	metricName := fmt.Sprintf("%s_info", strings.ToLower(kind))
+
+	return []ksmetric.FamilyGenerator{
+		ksmetric.FamilyGenerator{
+			Name: metricName,
+			Type: ksmetric.Gauge,
+			Help: helpText,
+			GenerateFunc: func(obj interface{}) *ksmetric.Family {
+				crd := obj.(*unstructured.Unstructured)
+				return &ksmetric.Family{
+					Metrics: []*ksmetric.Metric{
+						{
+							Value:       1,
+							LabelKeys:   []string{"namespace", kindName},
+							LabelValues: []string{crd.GetNamespace(), crd.GetName()},
+						},
+					},
+				}
+			},
+		},
+	}
+}
+
+// FilterOut takes in the operator specific scheme and filters out all generic apimachinery meta types.
+// It returns the GVK specific to this operator.
+func FilterOutMetaTypes(operatorSpecificScheme *runtime.Scheme) map[schema.GroupVersionKind]reflect.Type {
```

Comment:
WDYT of refactoring to something like this (maybe with a better function name :slightly_smiling_face:):

```go
func BaseGVKsFromAPIRegistration(addToSchemeFunc func(*runtime.Scheme) error ) ([]schema.GroupVersionKind, error) {
	s := kuberuntime.NewScheme()
	err := addToSchemeFunc(s)
	if err != nil {
		return nil, err
	}
	gvks := []schema.GroupVersionKind{}
	allKnownTypes := s.AllKnownTypes()
	for gvk, _ := range allKnownTypes {
		if !isMetaKind(gvk) {
			gvks = append(gvks, gvk)
		}
	}
	return gvks, nil
}
```

Then we wouldn't need the extra helper function in `main.go`. It would just look like:
```go
...
	// This returns the base GVKs that apis.AddToScheme registers to a scheme.
	// If you need more control, remove this and create your GVK list with your own custom logic.
	gvks, err := BaseGVKsFromAPIRegistration(apis.AddToScheme)
	if err != nil {
		log.Error(err, "Could not get base GVKs")
	}
	err := kubemetrics.ServeCRMetrics(cfg, []string{}, gvks, metricsHost, operatorMetricsPort)
	if err != nil {
		log.Error(err, "Could not serve Custom Resource metrics")
	}
...
```

Reply:
That sounds good to me, less logic in the users code is better :) 

> BaseGVKsFromAPIRegistration
WDYT of `GetOperatoSpecificGVKs` naming is not hard :)

Reply:
I thought about something like `GetOperatorSpecificGVKs()` but the function returns the GVKs from the provided `AddToScheme` function, which could be anything really. That's why I was trying to come up with something more generic.

Maybe `GetGVKsFromAddToScheme`?

Reply:
@joelanford Done, can you have a look again, thanks!